app/jobs/regular/emit_web_hook_event.rb

```diff
@@ -71,71 +52,164 @@ def web_hook_request(args, web_hook)
-          'X-Discourse-Event-Type' => args[:event_type]
-        }
-
-        headers['X-Discourse-Event'] = args[:event_name].to_s if args[:event_name].present?
-
-        if web_hook.secret.present?
-          headers['X-Discourse-Event-Signature'] = "sha256=#{OpenSSL::HMAC.hexdigest("sha256", web_hook.secret, body)}"
-        end
-
         now = Time.zone.now
-        response = conn.post(headers: headers, body: body)
-
+        response = conn.post(headers: web_hook_headers, body: web_hook_body)
         web_hook_event.update!(
-          headers: MultiJson.dump(headers),
+          headers: MultiJson.dump(web_hook_headers),
           status: response.status,
           response_headers: MultiJson.dump(response.headers),
           response_body: response.body,
           duration: ((Time.zone.now - now) * 1000).to_i
         )
       rescue => e
         web_hook_event.update!(
-          headers: MultiJson.dump(headers),
+          headers: MultiJson.dump(web_hook_headers),
           status: -1,
           response_headers: MultiJson.dump(error: e),
           duration: ((Time.zone.now - now) * 1000).to_i
         )
       end
 
+      publish_webhook_event(web_hook_event)
+      retry_web_hook if response&.status != 200
+    end
+
+    def retry_web_hook
+      if SiteSetting.retry_web_hook_events?
+        arguments[:retry_count] = (arguments[:retry_count] || 0) + 1
+        return if arguments[:retry_count] > MAX_RETRY_COUNT
+        delay = RETRY_BACKOFF**(arguments[:retry_count] - 1)
+        Jobs.enqueue_in(delay.minutes, :emit_web_hook_event, arguments)
+      end
+    end
+
+    def publish_webhook_event(web_hook_event)
       MessageBus.publish("/web_hook_events/#{web_hook.id}", {
         web_hook_event_id: web_hook_event.id,
-        event_type: args[:event_type]
+        event_type: event_type
       }, user_ids: User.human_users.staff.pluck(:id))
+    end
 
-      retry_web_hook if response&.status != 200
+    def ping_event?(event_type)
+      PING_EVENT == event_type.to_s
     end
 
-    def retry_web_hook
-      if SiteSetting.retry_web_hook_events?
-        @orig_args[:retry_count] = (@orig_args[:retry_count] || 0) + 1
-        return if @orig_args[:retry_count] > MAX_RETRY_COUNT
-        delay = RETRY_BACKOFF**(@orig_args[:retry_count] - 1)
-        Jobs.enqueue_in(delay.minutes, :emit_web_hook_event, @orig_args)
+    def webhook_inactive?
+      !web_hook.active?
+    end
+
+    def group_webhook_invalid?
+      web_hook.group_ids.present? && (group_id.present? ||
+        !web_hook.group_ids.include?(group_id))
+    end
+
+    def category_webhook_invalid?
+      web_hook.category_ids.present? && (!category_id.present? ||
+        !web_hook.category_ids.include?(category_id))
+    end
+
+    def tag_webhook_invalid?
+      web_hook.tag_ids.present? && (tag_ids.blank? ||
+        (web_hook.tag_ids & tag_ids).blank?)
+    end
+
+    def arguments
+      @arguments
+    end
+
+    def event_type
```

Comment:
I don't see the advantage of memoizing these values. Hash accesses are really fast. 

If it's to avoid calling `to_s` repeatedly, I would suggest adding @event_type = arguments[:event_type].to_s in the main `execute` method. However most of them are simply numbers and don't seem to gain anything by being memoized.

Reply:
> I don't see the advantage of memoizing these values. Hash accesses are really fast.

Makes sense, we don't think need memoization except for [this](https://github.com/discourse/discourse/pull/7385/files#diff-0e98636b29e3b908201e344e101066baR164).

> If it's to avoid calling to_s repeatedly, ....

The reason I wrap them into methods is that we are using them across different methods in this job, it feels better in terms of readability. At the first look at the previous version of the file, arguments are scattered all over the place.